### PR TITLE
Feature: Add afterDone hook to Compiler

### PR DESCRIPTION
**Pull request by mc-zone**
Add `afterDone` hook to Compiler which would be triggered only after other `done` hooks and `callbacks` finished.

Previous issue: #6825 
Hint: To resolve this, we should change other packages(like `webpack-dev-middleware`) to use this hook subsequently after release.

Also found in [webpack5 project board](https://github.com/webpack/webpack/projects/5#card-13643668)

**What kind of change does this PR introduce?**
feature.

**Did you add tests for your changes?**
Yes.

**Does this PR introduce a breaking change?**
No.

**What needs to be documented once your changes are merged?**
Should add more descriptions about compiler hook in docs.

### CODE REVIEW

File: lib/Compiler.js [hunk 1]
```diff
@@ -192,7 +194,12 @@ class Compiler {
 			this.cache.beginIdle();
 			this.running = false;
 
-			if (callback !== undefined) return callback(err, stats);
+			let callbackRst;
```
Comment (ooflorent) [inline lib/Compiler.js]: You don’t need this. `SyncHook` returns nothing

Comment (mc-zone) [reply lib/Compiler.js]: This `return` is for compatible with former behavior about the `finalCallback` return value. Has this useless before?

Comment (sokra) [reply lib/Compiler.js]: yes, we tend to use `if(...) return callback(...)` as shortcut for `if(...) { callback(...); return; }`

File: lib/Watching.js [hunk 2]
```diff
@@ -149,6 +149,9 @@ class Watching {
 			}
 			for (const cb of this.callbacks) cb();
 			this.callbacks.length = 0;
+			if (this.closed) {
```
Comment (sokra) [inline lib/Watching.js]: This can be removed. afterDone should be called after each run.

File: test/Compiler.test.js [hunk 3]
```diff
@@ -517,6 +517,49 @@ describe("Compiler", () => {
...
+			compiler.run((err, stats) => {
+				if (err) return done(err);
+				done();
+			});
+		});
+		compiler.run((err, stats) => {
+			if (err) return done(err);
+		});
+	});
+	it("should watch again correctly inside afterDone hook", function(done) {
```
Comment (sokra) [inline test/Compiler.test.js]: Not sure if the test makes sense this way

File: lib/Watching.js (hunk 2 above)
Comment (mc-zone) [reply lib/Watching.js]: Got it.

File: test/Compiler.test.js (hunk 3 above)
Comment (mc-zone) [reply test/Compiler.test.js]: I have changed tests to focus on the hook's call sequence.

**Comment by mc-zone** [comment]
@sokra Could you please review again?

**Comment by mc-zone** [comment]
Looking forward to hearing your advice :) @sokra @ooflorent

**Comment by sokra** [comment]
Thanks

**Comment by mc-zone** [comment]
ok! I will do it soon.


 
---Original---

**Pull request merged**